submissions/Roophee/pair-game/index.js

```diff
@@ -0,0 +1,173 @@
+        }
+    }
+}
+
+    //MAKE ANIMATION ON PAGE DOWNLOAD
+const animateOnStart = function(){
+    gameBg.style.top = `${-gameBg.clientHeight - gameBg.offsetTop}px`;
+    cardPlace.style.bottom = `${-document.documentElement.clientHeight * 2}px`;
+    setTimeout(function (){
+        gameBg.style.top = '0px';
+        cardPlace.style.bottom = '0px';
+    },500)
+}
+
+document.addEventListener('DOMContentLoaded', animateOnStart);
+
+// ADD CARDS ON DOCUMENT
+let addCardsOnPage = function () {
+    let cardsContainerOnDynamicGeneration = document.createDocumentFragment();
+    for (let i = 0; i < quantityOfCardsOnTheScreen; i++) {
+        let cardNameInsertOnDynamicGeneration = heroesArray[indexInUse[getRandomImage()]],
+            cardItem = document.createElement('div');
+        cardItem.classList.add('game__card-holder');
+        cardItem.innerHTML = `<div class="game__card-item">
+                            <div class="game__card-front">
+                                <img src="img/${cardNameInsertOnDynamicGeneration}.jpg" alt="${cardNameInsertOnDynamicGeneration}"  class="game__card-img">
+                            </div>
+                            <div class="game__card-back">
+                                <img src="img/cardback.jpg" alt="cardback" class="game__card-cardback">
+                            </div>
+                        </div>`;
+        cardsContainerOnDynamicGeneration.appendChild(cardItem);
+    }
+    cardPlace.appendChild(cardsContainerOnDynamicGeneration);
+}
+
+addCardsOnPage();
+
+//  CLEAR ARRAYS AND OBJECTS, PICK NEW CARDS FOR THE GAME
+let clearAllOperationObjectsAndCreateNewGame = function() {
+    if (pairs.length === maxQuantityOfPairsOnTheScreen) {
+        let playAgain = window.confirm('Congratulations. You have matched all pairs and won the game. Want to play again?');
+                if (playAgain) {
+                    cardPlace.innerHTML = "";
+                    for (const prop of Object.keys(imagesCountForTwo)) {
+                        delete imagesCountForTwo[prop];
+                        };
+                    indexInUse.length = 0;
+                    prePair.length = 0;
+                    pairs.length = 0;
+                    cardsForPlay();
+                    addCardsOnPage();
+                }
+            }
+}
+
+// CHECK PLAYER COLLECTED ALL PAIRS OR NOT
+const checkGameIsOverOrNot = function () {
+    setTimeout(function () {
+            clearAllOperationObjectsAndCreateNewGame();
+    }, 2100)
+}
+
+// ADD HIDE CLASS ON CARD-HOLDER ITEM IF TWO CARD ARE PAIR
+let hidePair = function(){
+    if (prePair[0] && prePair[1]){
+    prePair[0][1].classList.add('hide');
+    prePair[1][1].classList.add('hide');
+    prePair.splice(0, 2);
+    }
+}
+
+// REMOVE ACTIVE CLASS FROM CARD-HOLDER ITEM IF TWO CARD ARE NOT A PAIR
+let clearPair = function() {
+    prePair[0][1].classList.remove('active');
+    prePair[1][1].classList.remove('active');
```

Comment:
what stored here ? not readable 
```
prePair[0][1]
prePair[1][1]
```

Reply:
array pre-Pair  - is operational array.
When palyer open first card it added in prePair, then second card. When prePair.lenght = 2 it is the wright  momemt to check  , two card elements are pair or not. After this prePair clears up and ready for new cards from player.

How it fix better?

Reply:
1. My question was what stored exactly in this index  `prePair[0][1]`. It is not obvious for other developers, so they must search in your code what is going before this manipulations.

1. Honestly, i don't understand, why you need all of this variables. `heroesArray` - is array with main data, (but change naming, `heroesArray` is a bad idea - you show data type in the variable name and in JS it’s redundant). but why you need other ones ?

```
heroesArray = ['luke', 'vader', 'solo', 'ob1', 'palpatine', 'leia', 'r2d2-c3po', 'chewbacca', 'bad1', 'bad2', 'bad3', 'bad4', 'good1', 'good3', 'good4', 'robot1', 'yoda'],
    indexInUse = [],
    imagesCountForTwo = {},
    prePair = [],
    pairs = [],
```

I think you have made the data structure too complex to work with and understand, especially for other people. 
Feel free to ask me questions, i guess, together we can simplify your app. :smile:

Reply:
@Roophee

Reply:
1) Array of cards. There are 17 images , that mean that in every new game will be in use new card. It is more interensting, that  play whis same cards.
2) I pick 6 of cards, and write in `indexInUse` indexes of this card.
3)I need create pair of card whis same `img`  - when js create DOMElement, I append index of card in object as the key, and in value of this key - I count how many card whis this index (image) was appended in DOM.  If value of key === 2, js can't append third.  And this works whis `random` algorithm.
4) `prePair` - is operational array.
When player open first card it added in prePair, then second card. When prePair.lenght = 2 it is the wright momemt to check , two card elements are pair or not. After this prePair clears up and ready for new cards from player.
5) pair - count how many pairs payer find.  When two card checked, that they are pair - algorithm add  `1` in `pairs`. When `pairs.lenght===6` game is over. Algorithm check this  when every new pair  was finded.

Reply:
> what stored here ? not readable
> 
> ```
> prePair[0][1]
> prePair[1][1]
> ```

in prePair I push array with 2 elements - card-ID( from `img`  `alt`) and card-Node for apply styles.
Then I comparing `prePair[0][0]` and `prePair[[1][0]` - ID of cards.
Then I apply styles for `prePair[0][1]` and `prePair[[1][1]`